.github/workflows/main.yaml

```diff
@@ -0,0 +1,186 @@
+name: CI Workflow
+
+on:
+  push:
+    branches:
+      - master
+      - main
+  pull_request:
+    branches:
+      - master
+      - main
+      - develop
+  workflow_dispatch:
+
+permissions:
+  contents: read
+  pull-requests: write
+
+env:
+  COMPONENT_BUILDER: builder
+  COMPONENT_API: api
+  AWS_ACCOUNT_ID: ${{ secrets.AWS_ACCOUNT_ID }}
+  AWS_ECR_REGION: us-east-2
+
+jobs:
+
+  lint-golangci:
+    runs-on: ubuntu-latest
+    if: github.event_name == 'pull_request'
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+
+      - uses: actions/setup-go@v3
+        with:
+          go-version: ">=1.19.0"
+      - run: go version
+
+      - name: go fmt
+        run: |
+          go fmt ./...
+
+      - name: Lint Go code
+        uses: golangci/golangci-lint-action@v3
+        with:
+          version: latest
+
+  lint-dockerfile:
+    runs-on: ubuntu-latest
+    if: github.event_name == 'pull_request'
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+
+      - name: Lint builder Dockerfile
+        uses: hadolint/hadolint-action@v2.1.0
+        with:
+          dockerfile: ./dockerfile/builder/Dockerfile
+
+      - name: Lint api Dockerfile
+        uses: hadolint/hadolint-action@v2.1.0
+        with:
+          dockerfile: ./dockerfile/api/Dockerfile
+
+  install-go:
```

Comment:
Can we combine this job with the go fmt and linting job so that we don't need to repeat the checkout and setup go actions? 

Also same for the other jobs. What is the advantage to separating them vs combining them into one job?